File: 05_code.py
input_file = "05_input.txt"
test_input_file = "05_test_input.txt"
# input_file = test_input_file

def update_is_valid(update, rules):
    valid_update = True
    for j, number in enumerate(update):
        left_of_number = update[:j]
        right_of_number = update[j+1:]

        left_rules = [rule[0] for rule in rules if number == rule[1]]
        right_rules = [rule[1] for rule in rules if number == rule[0]]

        left_violation = any([left_number in right_rules for left_number in left_of_number])
        right_violation = any([right_number in left_rules for right_number in right_of_number])

        violation = left_violation or right_violation

        if violation:
            valid_update = False

    return valid_update

rules = []
updates = []
with open(input_file, "r") as f:
    for line in f.readlines():
        if "|" in line:
            rules.append([int(number) for number in line.split("|") if number != "\n"])
        elif len(line.strip()) == 0:
            continue
        else:
            updates.append([int(number) for number in line.split(",") if number != "\n"])

invalid_updates = []
middle_number_of_valid_updates = []
for i, update in enumerate(updates):

    if update_is_valid(update, rules):
        assert len(update) % 2 == 1
        k = int(len(update) // 2)
        middle_number = update[k]
        middle_number_of_valid_updates.append(middle_number)
    else:
        invalid_updates.append(update)

print(f"First part: {sum(middle_number_of_valid_updates)}")


corrected_updates = []
for i, invalid_update in enumerate(invalid_updates):
    update = []
    while len(invalid_update):
        test_number = invalid_update.pop()
        for j in range(len(update)+1):
            tmp_update = update.copy()
            tmp_update.insert(j, test_number)
            if update_is_valid(tmp_update, rules):
                update = tmp_update
                break
    corrected_updates.append(update)

total_sum = 0
for update in corrected_updates:
    assert len(update) % 2 == 1
    k = int(len(update) // 2)
    middle_number = update[k]
    total_sum += middle_number

print(f"Second part: {total_sum}")

File: 05_code.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>

bool check_subset_for_violation(const std::vector<int> &numbers, const std::vector<int> &rules) {
    for (const int number : numbers) {
        for (const int rule : rules) {
            if (number == rule) {
                return true;
            }
        }
    }
    return false;
}


bool update_is_valid(const std::vector<std::pair<int, int>> &rules, const std::vector<int> &update) {
    bool is_valid = true;

    for (int i = 0; i < update.size(); i++) {
        const int number = update[i];

        std::vector<int> left_of_number(update.begin(), update.begin() + i);
        std::vector<int> right_of_number(
            std::next(
                update.begin(),
                std::min(static_cast<std::size_t>(i + 1), update.size())
                ),
            update.end());

        std::vector<int> left_rules;
        std::vector<int> right_rules;
        for (auto [fst, snd] : rules) {
            if (snd == number) {
                left_rules.push_back(fst);
            } else if (fst == number) {
                right_rules.push_back(snd);
            }
        }

        const bool left_violation = check_subset_for_violation(left_of_number, right_rules);
        const bool right_violation = check_subset_for_violation(right_of_number, left_rules);

        if (left_violation || right_violation) is_valid = false;
    }
    return is_valid;
}

int main() {

    const std::string filename = "05_input.txt";
    std::ifstream file(filename, std::ios::in);

    if (!file.is_open()) {
        std::cerr << "Error opening file " << filename << std::endl;
        return 1;
    }

    std::vector<std::pair<int, int>> rules;
    std::vector<std::vector<int>> updates;

    std::string line;
    while (std::getline(file, line)) {
        if (line.contains('|')) {

            std::size_t separator = line.find('|');
            int left_number = std::stoi(line.substr(0, separator));
            int right_number = std::stoi(line.substr(separator + 1));
            rules.emplace_back(left_number, right_number);

        } else if (line.contains(',')) {
            std::vector<int> numbers;
            std::string s_number;
            std::stringstream ss(line);
            while (std::getline(ss, s_number, ',')) {
                numbers.emplace_back(std::stoi(s_number));
            }
            updates.push_back(numbers);
        }
    }

    std::vector<std::vector<int>> invalid_updates;
    int first_part_sum = 0;
    for (std::vector<int> update : updates) {
        if (update_is_valid(rules, update)) {
            int middle_index = static_cast<int>(update.size() / 2);
            first_part_sum += update[middle_index];
        } else {
            invalid_updates.push_back(std::move(update));
        }
    }
    std::cout << "First part: " << first_part_sum << std::endl;

    int second_part_sum = 0;
    for (std::vector<int> invalid_update : invalid_updates) {
        std::vector<int> corrected_update;
        corrected_update.reserve(invalid_update.size());

        while (!invalid_update.empty()) {
            int test_number = invalid_update.back();
            invalid_update.pop_back();

            for (int i = 0; i < corrected_update.size() + 1; i++) {
                std::vector<int> tmp_update = corrected_update;
                tmp_update.insert(tmp_update.begin() + i, test_number);
                if (update_is_valid(rules, tmp_update)) {
                    corrected_update = tmp_update;
                    break;
                }
            }
        }

        int middle_index = static_cast<int>(corrected_update.size() / 2);
        second_part_sum += corrected_update[middle_index];
    }

    std::cout << "Second part: " << second_part_sum << std::endl;

    return 0;
}